`manual_unwrap_or_default` suggested when `ref` keyword is used

### Summary

The lint `manual_unwrap_or_default` is only suggested in the version where the `ref` keyword is used.


### Lint Name

manual_unwrap_or_default

### Reproducer

**I tried this code:**

```rust
struct Project {
    name: Option<String>,
}

fn func(project: &Project) {
    // this version triggers the lint
    let _name = match project.name {
        Some(ref x) => x,
        None => "",
    };

    // counter-example: this version does not trigger the lint
    let _name2 = match &project.name {
        Some(x) => x,
        None => "",
    };
}
```

**I saw this happen:**

```
error: match can be simplified with `.unwrap_or_default()`
   --> src/main.rs:156:17
    |
156 |       let _name = match project.name {
    |  _________________^
157 | |         Some(ref x) => x,
158 | |         None => "",
159 | |     };
    | |_____^ help: replace it with: `project.name.unwrap_or_default()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_unwrap_or_default
    = note: `-D clippy::manual-unwrap-or-default` implied by `-D clippy::all`
    = help: to override `-D clippy::all` add `#[allow(clippy::manual_unwrap_or_default)]`
```

Applying the suggestion results in an error:

```
error[E0507]: cannot move out of `project.name` which is behind a shared reference
    --> src/main.rs:156:22
     |
156  |     let _incorrect = project.name.unwrap_or_default();
     |                      ^^^^^^^^^^^^ ------------------- `project.name` moved due to this method call
     |                      |
     |                      help: consider calling `.as_ref()` or `.as_mut()` to borrow the type's contents
     |                      move occurs because `project.name` has type `std::option::Option<std::string::String>`, which does not implement the `Copy` trait
     |
note: `std::option::Option::<T>::unwrap_or_default` takes ownership of the receiver `self`, which moves `project.name`
    --> /Users/mrclmh/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/option.rs:1048:30
     |
1048 |     pub fn unwrap_or_default(self) -> T
     |                              ^^^^
help: you can `clone` the value and consume it, but this might not be your desired behavior
     |
156  |     let _incorrect = <std::option::Option<std::string::String> as Clone>::clone(&project.name).unwrap_or_default();
     |                      ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++            +
help: consider cloning the value if the performance cost is acceptable
     |
156  |     let _incorrect = project.name.clone().unwrap_or_default();
     |                                  ++++++++
```

**I expected to see this happen:**
The code should not trigger the lint, because the suggested code is not valid.


### Version

```text
rustc 1.85.0 (4d91de4e4 2025-02-17)
binary: rustc
commit-hash: 4d91de4e48198da2e33413efdcd9cd2cc0c46688
commit-date: 2025-02-17
host: aarch64-apple-darwin
release: 1.85.0
LLVM version: 19.1.7
```

### Additional Labels

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"profetia"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/clippy_lints/src/matches/manual_unwrap_or.rs
+++ b/clippy_lints/src/matches/manual_unwrap_or.rs
@@ -1,5 +1,6 @@
 use clippy_utils::consts::ConstEvalCtxt;
 use clippy_utils::source::{SpanRangeExt as _, indent_of, reindent_multiline};
+use rustc_ast::{BindingMode, ByRef};
 use rustc_errors::Applicability;
 use rustc_hir::def::Res;
 use rustc_hir::{Arm, Expr, ExprKind, HirId, LangItem, Pat, PatExpr, PatExprKind, PatKind, QPath};
@@ -16,7 +17,7 @@ use super::{MANUAL_UNWRAP_OR, MANUAL_UNWRAP_OR_DEFAULT};
 
 fn get_some(cx: &LateContext<'_>, pat: &Pat<'_>) -> Option<HirId> {
     if let PatKind::TupleStruct(QPath::Resolved(_, path), &[pat], _) = pat.kind
-        && let PatKind::Binding(_, pat_id, _, _) = pat.kind
+        && let PatKind::Binding(BindingMode(ByRef::No, _), pat_id, _, _) = pat.kind
         && let Some(def_id) = path.res.opt_def_id()
         // Since it comes from a pattern binding, we need to get the parent to actually match
         // against it.

--- a/tests/ui/manual_unwrap_or_default.fixed
+++ b/tests/ui/manual_unwrap_or_default.fixed
@@ -106,3 +106,16 @@ fn issue_12928() {
 fn allowed_manual_unwrap_or_zero() -> u32 {
     Some(42).unwrap_or_default()
 }
+
+mod issue14716 {
+    struct Foo {
+        name: Option<String>,
+    }
+
+    fn bar(project: &Foo) {
+        let _name = match project.name {
+            Some(ref x) => x,
+            None => "",
+        };
+    }
+}

--- a/tests/ui/manual_unwrap_or_default.rs
+++ b/tests/ui/manual_unwrap_or_default.rs
@@ -147,3 +147,16 @@ fn allowed_manual_unwrap_or_zero() -> u32 {
         0
     }
 }
+
+mod issue14716 {
+    struct Foo {
+        name: Option<String>,
+    }
+
+    fn bar(project: &Foo) {
+        let _name = match project.name {
+            Some(ref x) => x,
+            None => "",
+        };
+    }
+}

